src/vivarium/component.py

```diff
@@ -671,32 +673,49 @@ def get_data(
-            try:
-                if module == "self":
-                    data_getter = getattr(self, method)
-                else:
-                    data_getter = getattr(import_module(module), method)
-            except ModuleNotFoundError:
-                raise ConfigurationError(f"Unable to find module '{module}'.")
-            except AttributeError:
-                module_string = (
-                    f"component {self.name}." if module == "self" else f"module '{module}'."
-                )
-                raise ConfigurationError(
-                    f"There is no method '{method}' for the {module_string}."
-                )
-
-            return data_getter(builder)
-
-        try:
-            return builder.data.load(data_source)
-        except ArtifactException:
-            raise ConfigurationError(f"Failed to find key '{data_source}' in artifact.")
+        # TODO update this to use vivarium.types.LookupTableData once we drop
+        #  support for Python 3.9
+        valid_data_types = (Number, timedelta, datetime, pd.DataFrame, list, tuple)
+        if isinstance(data_source, valid_data_types):
+            data = data_source
+        elif isinstance(data_source, str):
+            if "::" in data_source:
+                module, method = data_source.split("::")
+                try:
+                    if module == "self":
+                        data_source = getattr(self, method)
+                    else:
+                        data_source = getattr(import_module(module), method)
+                except ModuleNotFoundError:
+                    raise ConfigurationError(f"Unable to find module '{module}'.")
+                except AttributeError:
+                    module_string = (
+                        f"component {self.name}" if module == "self" else f"module '{module}'"
+                    )
+                    raise ConfigurationError(
+                        f"There is no method '{method}' for the {module_string}."
+                    )
+                data = data_source(builder)
+            else:
+                try:
+                    data = builder.data.load(data_source)
+                except ArtifactException:
+                    raise ConfigurationError(
+                        f"Failed to find key '{data_source}' in artifact."
+                    )
+        elif isinstance(data_source, Callable):
+            data = data_source(builder)
+        else:
+            raise ConfigurationError(
+                f"Data source '{data_source}' is not a valid data source. It "
+                f"must be a LookupTableData instance, a string corresponding to "
+                f"an artifact key, a callable that returns a LookupTableData "
+                f"instance, or a string defining such a callable."
+            )
+
+        if not isinstance(data, valid_data_types):
+            raise ConfigurationError(f"Data '{data}' must be a LookupTableData instance.")
```

Comment:
nit: would it be better to print the `type(data)` to prevent it blowing up the log and not being helpful?